In [1]:
import pandas as pd

# The Palmer Archipelago Penguin Data

<img src="cute_penguins.png" alt="Drawing" style="width: 500px;"/>

Art by @allison_horst

In [2]:
# load the data
path = 'https://raw.githubusercontent.com/um-perez-alvaro/Data-Science-Theory/master/Data/penguins_size.csv'
df = pd.read_csv(path)
df

species     island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen              39.1             18.7              181.0   
1    Adelie  Torgersen              39.5             17.4              186.0   
2    Adelie  Torgersen              40.3             18.0              195.0   
3    Adelie  Torgersen               NaN              NaN                NaN   
4    Adelie  Torgersen              36.7             19.3              193.0   
..      ...        ...               ...              ...                ...   
339  Gentoo     Biscoe               NaN              NaN                NaN   
340  Gentoo     Biscoe              46.8             14.3              215.0   
341  Gentoo     Biscoe              50.4             15.7              222.0   
342  Gentoo     Biscoe              45.2             14.8              212.0   
343  Gentoo     Biscoe              49.9             16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  
1         3800.0  FEMALE  
2         3250.0  FEMALE  
3            NaN     NaN  
4         3450.0  FEMALE  
..           ...     ...  
339          NaN     NaN  
340       4850.0  FEMALE  
341       5750.0    MALE  
342       5200.0  FEMALE  
343       5400.0    MALE  

[344 rows x 7 columns]

Data were collected and made available by Dr. Kristen Gorman and the Palmer Station, Antarctica LTER, a member of the Long Term Ecological Research Network.
This dataset contains data for 344 penguins.
There are 3 different species of penguins in this dataset, collected from 3 islands in the Palmer Archipelago, Antarctica.

In [4]:
df.species.value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

The culmen is the upper ridge of a bird’s bill. For this penguin data, the culmen (bill) length and depth are measured as shown below.

<img src="culmen_depth.png" alt="Drawing" style="width: 500px;"/>

The **goal** is to predict the penguin species from physical measurements (culmen length, culmen depth, flipper length, and body mass). 

**Part 1:** Import and instantiate a k-nearest neighbors model.

In [ ]:
# your  code here


**Part 2:** Define the feature matrix X and the target vector y from the dataframe, and then split X and y into training and testing sets.

In [ ]:
# your  code here


**Part 3:** Build a pipeline that consists of two steps: an imputer (to fill missing values with the mean value) and the knn classifier.

In [ ]:
# your  code here


**Part 4:** Use a grid search to tune in the `n_neighbors` and `weights` hyperparameters

<div class="admonition note alert alert-info">
<p class="first admonition-title" style="font-weight: bold;">Note</p>
<p>  The Palmer Archipelago Penguin Data is relatively small.
    Don't use a large value for the <tt> cv </tt> parameter</p> (I would use <tt> cv=5</tt>.)
</div>

In [ ]:
# your  code here


**Part 5:** Use accuracy and the confusion matrix to evaluate the performance of your best model on the test set

In [ ]:
# your  code here


In [ ]:
# accuracy


In [ ]:
# confusion matrix


<div class="alert alert-block alert-danger"> <tt>KNeighborsClassifier </tt> is a distance based model. Distance based models are affected by the scale of the features. They give higher weightage to features which have higher magnitude (The <tt>body_mass_g</tt> feature in our case).
We do not want our classifier to be affected by the magnitude of the features. To overcome this problem, we can bring down all the variables to the same scale. </div>

**Part 6:** Add a `StandardScaler` to your pipeline, and repeat steps 4 and 5.

In [ ]:
# your  code here
